### ISSUE TITLE
Linking to Repos In-App 

### ISSUE THREAD

**Issue by BrianLitwin** [GitHawkApp/GitHawk#2247]
Do you all think it would be a good idea to catch links to repos in repository ReadMe's and open a RepositoryViewController instead of re-directing the user to Github.com? Eg when a user clicks the Apollo Codegen link in GitHawk's ReadMe?  There's already a workflow in place that we use for shortlinks, seems like it would be a matter of writing a regex to catch links to repos and wiring it up.

**Comment by BrianLitwin** [comment]
for further clarification: 

instead of opening repo links in safari: 
![ezgif com-video-to-gif 2](https://user-images.githubusercontent.com/26695477/46749908-c0f79300-cc84-11e8-9cce-545cd7e17d35.gif)


we could open the links in GitHawk: 
![ezgif com-video-to-gif 3](https://user-images.githubusercontent.com/26695477/46750007-01efa780-cc85-11e8-88ba-a9f5fe644f9b.gif)

**Comment by rnystrom** [comment]
@BrianLitwin yes! that would be great

**Comment by BrianLitwin** [comment]
@rnystrom @Huddie After the user clicks a repository link, we gotta go grab the defaultBranchRef + hasIssuesEnabled info from graphql before opening the RepositoryViewController... I'm currently throwing a spinner here... lmk y'alls thoughts or if you have a better idea of how to keep the UI interactive while this quick fetch happens --> 


![ezgif com-video-to-gif 4](https://user-images.githubusercontent.com/26695477/46916974-d2e07b00-cf8f-11e8-9756-e4f8e1f7fbdf.gif)



<img width="296" alt="screenshot 2018-10-14 08 54 03" src="https://user-images.githubusercontent.com/26695477/46916982-ee4b8600-cf8f-11e8-8efb-fb51b74facfa.png">

**Comment by Huddie** [comment]
Cool stuff! The spinner seems different from the regular one which Idk how I feel about. I get that throwing a spinner on the view could get lost in the text and that’s why you prob opted to go with a spinner with a background view behind it. 
Maybe open repositoryViewController right away and have it load then? So it’s empty with the regular spinner rather than waiting on the previous page for the load.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnystrom** [comment]
Ya I almost think adding the ability to init a new repo VC without a branch would be good. Make it optional and the VC fetches it?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Huddie** [comment]
^^ yup

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BrianLitwin** [comment]
@Huddie  @rnystrom .. I'm fine with that.. only caveat is that after fetching the graphql info, graphql could return an error or "repo not found" and the process would like: 
- user taps link 
- segue to empty repo view controller 
- fetch repo info 
- fetch fails 
- segue to safari view controller 

An example of that is with our Apollo-codegen link.. the repo was renamed and GraphQL responds with "repo not found" 

I think the empty vc approach is still the right move, just have to make it seamless if loading the repo fails.

**Comment by j-f1** [comment]
You could present the Safari VC as a modal with a fade, then pop the repo view controller without animation once the Safari VC is completely visible.

**Comment by Huddie** [comment]
🤔 maybe an alert saying “repo not found, try opening in web” with a button that when clicked will redirect you? Or I guess you could just redirect them 🤷‍♂️ 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnystrom** [comment]
@BrianLitwin so V3 handles the rename but GraphQL doesn't?

**Comment by BrianLitwin** [comment]
@rnystrom yes that seems to be the case that it works w/ v3 but not GraphQL.

**Comment by rnystrom** [comment]
I'll try to file an issue w/ GitHub on this

**Comment by BrianLitwin** [comment]
Haha @rnystrom  if you/they want a test case to copy/paste in the [GraphQL Explorer](https://developer.github.com/v4/explorer/): 

```graphql
query($owner:String!, $name:String!) {
repository(owner:$owner, name:$name) {
    hasIssuesEnabled
  }
}
```
```graphql
{ "owner": "apollographql", "name": "apollo-codegen" }
```

**Comment by BrianLitwin** [comment]
just to track my thinking/progress on this: 

the repoDetails object that a RepoViewController needs to initialize is: 

```swift
struct RepositoryDetails: Codable, Equatable {
    let owner: String
    let name: String
    let defaultBranch: String
    let hasIssuesEnabled: Bool
}
```

If we recognize a url that looks like a repository url in the markdown of a README, we can get `owner` and `name`. Now we want to push a `RepositoryViewController` instead of linking to a safari webview. 

`RepositoryViewController` currently looks like: 

```swift 
    private let repo: RepositoryDetails
    private let client: GithubClient
    private let controllers: [UIViewController] 
    private var bookmarkNavController: BookmarkNavigationController?
    public private(set) var branch: String

    init(client: GithubClient, repo: RepositoryDetails) { ... }
 
```

> Ya I almost think adding the ability to init a new repo VC without a branch would be good. Make it optional and the VC fetches it?

So I'm hesitant to use a failable initializer/ fetching `defaultBranch` and `hasIssuesEnabled` from RepoVC because we'd have to change `let repo: RepositoryDetails` to `var repo: RepositoryDetails?'` and make `let controllers` a `var` too (since it requires hasIssuesEnabled to know whether to include an IssuesVC). That feels like a compromise that out-weighs the benefits. 

I haven't had any luck tho pushing an empty VC w/ a loading spinner and then pushing the RepoVC on top of it with no animation or a fade-in. It looks okay but not sharp.
